src/net/sf/jsignpdf/SignerLogic.java

```diff
@@ -132,9 +143,67 @@ public boolean signFile() {
 		try {
 			SSLInitializer.init(options);
 
-			final PrivateKeyInfo pkInfo = KeyStoreUtils.getPkInfo(options);
-			final PrivateKey key = pkInfo.getKey();
-			final Certificate[] chain = pkInfo.getChain();
+			final PrivateKeyInfo pkInfo;
+			final PrivateKey key;
+			final Certificate[] chain;
+			// we first load certificates
+			if (this.options.getKsType().equals("remote")){
+				String remoteAddress = this.options.getKsFile();
+				if ((remoteAddress == null) || (!remoteAddress.contains(":")) || (this.options.getKeyAlias() == null)) {
+					return false;
+				}
+				key = null;  // the private key is inside the smart card, we don't have it here
+
+				int cmdId = (int)(Math.random() * 100000);
+				String cert_chain_request = MessageFormat.format(">{0}|\n>{1}:CHAIN|",
+						this.options.getKeyAlias(), cmdId);
+
+				String address [] = options.getKsFile().split(":");
+				String hostname = address[0];
+				int port = Integer.parseInt(address[1]);
+
+				try {
+					Socket socket = new Socket(hostname, port);
+
+					OutputStream output = socket.getOutputStream();
+					PrintWriter writer = new PrintWriter(output, true);
+					writer.println(cert_chain_request);
+
+					InputStream input = socket.getInputStream();
+
+					BufferedReader reader_sock = new BufferedReader(new InputStreamReader(input));
+
+					String line = reader_sock.readLine();
+
+					socket.close();
+
+					if (line == null) {
+						return false;
+					}
+					String []certChain = line.split(":");
+					chain = new Certificate[certChain.length - 1];
+					CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+					InputStream in;
+					for (int i = 1; i<certChain.length; i++) {
+						// convert the certificate
+						BASE64Decoder decoder = new BASE64Decoder();
+						byte[] decodedBytes = decoder.decodeBuffer(certChain[i]);
+						in = new ByteArrayInputStream(decodedBytes);
+						chain[i - 1] = certFactory.generateCertificate(in);
+					}
+				} catch (UnknownHostException ex) {
+					LOGGER.error("Can't connect to signing server" + hostname +": " + ex.getMessage());
```

Comment:
The messages should be handled through resources, so the translations are possible.